Raise NotImplementedError if using wrong class

vonschultz opened the pull request:
Instead of leaving abstract methods without definition when running on the other platform (Unix vs Windows), define them and raise a `NotImplementedError` if called on the wrong platform.

This fixes pylint warnings such as
```
   Abstract class 'WindowsFileLock' with abstract methods instantiated
   Abstract class 'UnixFileLock' with abstract methods instantiated
```

Fixes tox-dev/py-filelock#102

vonschultz committed:
Raise PlatformMismatchError if using wrong class

Instead of leaving abstract methods without definition when running on
the other platform (Unix vs Windows), define them and raise a
PlatformMismatchError if called on the wrong platform.

This fixes pylint warnings such as
   Abstract class 'WindowsFileLock' with abstract methods instantiated
   Abstract class 'UnixFileLock' with abstract methods instantiated

Fixes tox-dev/py-filelock#102

vonschultz committed:
Remove unused ABC import

gaborbernat reviewed (changes requested):
I prefer in this case the class to remain abstract. This is a clearly faulty warning in pylint and the solution should be to change pylint check than modify our code. #102 was closed for a reason: the issue is not with us, is with pylint.

gaborbernat commented on src/filelock/_api.py:
```diff
@@ -14,6 +14,10 @@
 _LOGGER = logging.getLogger("filelock")
 
 
+class PlatformMismatchError(Exception):
```
Let's remove this class 👍

gaborbernat commented on src/filelock/_unix.py:
```diff
@@ -2,18 +2,23 @@
         """Uses the :func:`fcntl.flock` to hard lock the lock file on unix systems."""
 
+        def _acquire(self) -> None:
+            raise PlatformMismatchError()
```
Let's just change this and the others

```raise NotImplementedError```

gaborbernat reviewed (changes requested):
You'll want to add tests too 👍  and a changelog entry.

vonschultz committed:
Use NotImplementedError, not PlatformMismatchError

Instead of defining a custom exception, PlatformMismatchError, use the
standard NotImplementedError.

vonschultz committed:
Add tests and a note in the change log

Add tests of UnixFileLock on Windows systems and WindowsFileLock on
non-Windows systems, to make sure they are not abstract (but still
raise NotImplementedError). This avoids pylint errors in code using
filelock.

Also added a note to the change log.

gaborbernat commented on docs/changelog.rst:
```diff
@@ -1,6 +1,13 @@
 Changelog
 =========
 
+Future
```
Mark it the next minor and today's date 👍

vonschultz committed:
Merge remote-tracking branch 'upstream/main' into platform-mismatch-error

gaborbernat commented on tests/test_filelock.py:
```diff
@@ -368,3 +369,35 @@ def test_poll_intervall_deprecated(lock_type: type[BaseFileLock], tmp_path: Path
+        lock.acquire()
+
+    with pytest.raises(NotImplementedError):
+        lock._release()
```
Can you make this instead one parametrized test (based on platform)

vonschultz replied:
Do you mean with `@pytest.mark.parametrize()`? I don't really see how that would be done here. I can make it a single test fairly easily, though.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

vonschultz committed:
Tell flake8 that isabstract is not a misspelling

The Python standard library module "inspect" has a function called
"isabstract". That is not a spelling mistake. Silence flake8 SC200
warning about "isabstract".

vonschultz committed:
Make the wrong-platform tests a single test

Make a single test, and select lock_type based on sys.platform.

vonschultz committed:
Set release date to today

Next minor release, and today's release date in the change log.

gaborbernat commented on tests/test_filelock.py:
```diff
@@ -385,26 +385,13 @@ def decorated_method() -> None:
+    lock_type = UnixFileLock if sys.platform == "win32" else WindowsFileLock
+    lock = lock_type(str(tmp_path / "lockfile"))
 
     assert not inspect.isabstract(  # noqa: SC200
```
a better test here would be to instantiate the class and test that calling the acquire/release raises NotImplementedError 🤔

vonschultz replied:
I do instantiate the class and test that acquire/release raises NotImplementedError.

vonschultz replied:
We could skip the isabstract() part entirely, since it's implicit in that we instantiate the class. Though saying it explicitly makes it clearer that the non-abstractness is a desired outcome in itself, and is being tested on purpose.

gaborbernat replied:
sure you can do that too, but then still need to instantiate to check that calling those methods raises; up to you; and yes I'd make it parametrized personally, but you can get by with if statements too

vonschultz committed:
Move isabstract check before lock instantiation

To make it a bit clearer, move the preliminary check of abstractness
before the instantiation of a lock, so that the creation of the lock
and the pytest.raises(NotImplementedError) are adjacent.

gaborbernat committed:
PR feedback

Signed-off-by: Bernát Gábor <bgabor8@bloomberg.net>
